tasks/Task.cpp

```diff
@@ -2,48 +2,422 @@
+        });
+
+    mWs->onMessage(
+        [&](variant<binary, string> data)
+        {
+            if (!holds_alternative<string>(data))
+            {
+                return;
+            }
+
+            parseIncomingMessage(get<string>(data).c_str());
+            string actiontype = mDecoder.getActionType();
+
+            if (actiontype == "ping")
+            {
+                pong();
+            }
+            if (actiontype == "pong")
+            {
+                if (_remote_peer_id.get() == mDecoder.getFrom())
+                {
+                    mRemotePeerAnswerReceived = true;
+                    mWaitRemotePeerPromise.set_value();
+                }
+            }
+
+            if (actiontype == "offer")
+            {
+                onOffer();
+            }
+            else if (actiontype == "answer")
+            {
+                onAnswer();
+            }
+            else if (actiontype == "candidate")
+            {
+                onCandidate();
+            }
+        });
+
+    // wss://signalserverhost?user=yourname
+    const string url = _signaling_server_name.get() + "?user=" + _local_peer_id.get();
+    mWs->open(url);
+    ws_future.get();
+}
+
+void Task::ping()
+{
+    Json::Value message;
+    message["protocol"] = "one-to-one";
+    message["to"] = _remote_peer_id.get();
+    message["action"] = "ping";
+    message["data"]["from"] = _local_peer_id.get();
+    Json::FastWriter fast;
+    if (auto ws = make_weak_ptr(mWs).lock())
+    {
+        ws->send(fast.write(message));
+    }
+}
+
+void Task::pong()
+{
+    Json::Value message;
+    message["protocol"] = "one-to-one";
+    message["to"] = mDecoder.getFrom();
+    message["action"] = "pong";
+    message["data"]["from"] = _local_peer_id.get();
+    Json::FastWriter fast;
+    if (auto ws = make_weak_ptr(mWs).lock())
+    {
+        ws->send(fast.write(message));
+    }
+}
+
+Task::Task(string const &name) : TaskBase(name) {}
+
+Task::~Task() {}
 
 /// The following lines are template definitions for the various state machine
 // hooks defined by Orocos::RTT. See Task.hpp for more detailed
 // documentation about them.
 
 bool Task::configureHook()
 {
-    if (! TaskBase::configureHook())
+    if (!TaskBase::configureHook())
         return false;
+
+
+    mConfig.iceServers.emplace_back(_stun_server.get());
+    mWs = std::make_shared<rtc::WebSocket>();
+
+    configureWebSocket();
+
     return true;
 }
 bool Task::startHook()
 {
-    if (! TaskBase::startHook())
+    if (!TaskBase::startHook())
         return false;
+
+    std::future<void> dc_future = mDataChannelPromise.get_future();
+    std::future<void> wait_remote_peer_future = mWaitRemotePeerPromise.get_future();
+
+    if (!_remote_peer_id.get().empty())
+    {
+        mRemotePeerAnswerReceived = false;
+        while(!mRemotePeerAnswerReceived)
+        {
+            ping();
+            usleep(100);
```

Comment:
Waaaaay too quick. Should be 100ms (usleep is in microseconds).

Also, use the C++ sleep API

Reply:
ok